src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingSourceGenerator.cs

```diff
@@ -0,0 +1,194 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+//#define LAUNCH_DEBUGGER
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.DotnetRuntime.Extensions;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
+{
+    /// <summary>
+    /// Generates source code to optimize binding with ConfigurationBinder.
+    /// </summary>
+    [Generator]
+    public sealed partial class ConfigurationBindingSourceGenerator : IIncrementalGenerator
+    {
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            IncrementalValueProvider<KnownTypeData> compilationData =
+                context.CompilationProvider
+                    .Select((compilation, _) => new KnownTypeData(compilation));
+
+            IncrementalValuesProvider<BinderInvocationOperation> inputCalls = context.SyntaxProvider.CreateSyntaxProvider(
+                (node, _) => node is InvocationExpressionSyntax invocation,
+                (context, cancellationToken) => new BinderInvocationOperation(context, cancellationToken));
+
+            IncrementalValueProvider<(KnownTypeData, ImmutableArray<BinderInvocationOperation>)> inputData = compilationData.Combine(inputCalls.Collect());
+
+            context.RegisterSourceOutput(inputData, (spc, source) => Execute(source.Item1, source.Item2, spc));
+        }
+
+        /// <summary>
+        /// Generates source code to optimize binding with ConfigurationBinder.
+        /// </summary>
+        private static void Execute(KnownTypeData typeData, ImmutableArray<BinderInvocationOperation> inputCalls, SourceProductionContext context)
+        {
+#if LAUNCH_DEBUGGER
+            #pragma warning disable IDE0055
+            if (!System.Diagnostics.Debugger.IsAttached)
+            {
+                System.Diagnostics.Debugger.Launch();
+            }
+            try
+            {
+#endif
+            if (inputCalls.IsDefaultOrEmpty)
+            {
+                return;
+            }
+
+            Parser parser = new(context, typeData);
+            SourceGenerationSpec? spec = parser.GetSourceGenerationSpec(inputCalls);
+            if (spec is not null)
+            {
+                Emitter emitter = new(context, spec);
+                emitter.Emit();
+            }
+#if LAUNCH_DEBUGGER
+            }
+            catch (System.Exception ex)
+            {
+                System.Diagnostics.Debugger.Break();
+                throw ex;
+            }
+            #pragma warning restore
+#endif
+        }
+
+        private readonly record struct KnownTypeData
```

Comment:
Does this need to be a struct? It's pretty large so passing it by value would be pretty expensive.